umbraco/Umbraco-CMS
# Starts getting front-end routing/rendering working

This will be the first step in getting the front-end routing/rendering working. I've added a ton of TODOs a lot of these will need to be cleaned up on the 2nd iteration but should be looked at and commented on if necessary as part of the review.

* PublishedSnapshotService - split up
  * Repository for DB interaction
    * Changes to repo base class to make things easier, RepositoryBase (for all repos) + EntityRepositoryBase (for IEntity based repos) and no more NPocoRepositoryBase
  * Service for repository interaction - INuCacheContentService replaces IDataService
  * PublishedSnapshotServiceEventHandler for wiring events for PublishedSnapshotService
  * Remove EnterPreview/RefreshPreview/ExitPreview since this was only used for the xml cache
  * Still work in progress, there's still temporal coupling with `LoadCachesOnStartup`, how to avoid this?
* UmbracoRouteValueTransformer for handling the dynamic routes - will basically replace RenderRouteHandler and the module
* ContentModelBinder fixed up and tests for it combined into one suite (there was 2x test classes for the same thing)
* UmbracoViewPage fixed up so that the dynamic model binding works
* Fixup the startup ordering for UseUmbraco, etc... UseRouting is what installs endpoint routing so any middleware after that will not execute before dynamic routes in endpoint routing even if they are declared before UseEndpoints. However some things must be declared between UseRouting and UseEndpoints like AuthN, AuthZ. Code has also been updated to cater better for enabling front-end vs back office 

Others

* Remove IPublishedContentType2 backward compat hack
* Remove other unnecessary things
* Linting cleanup

This POC will not contain surface controller routes, post backs, forms, content finder redirections, route hijacking, custom routes, macros, etc... just the first step of rendering something

Further PRs for front-end routing will be based on this work

## testing

* Ensure tests are passing
* Create a content type with a template, let's say it's called "Home", that should create a view under ~/Views/Home.cshtml
* put some simple markup in there like (basically render a menu system and the name)
   ```
    @using Umbraco.Web.PublishedModels;
    @using Umbraco.Core;
    @using Umbraco.Web.Routing; 
    @inject IPublishedUrlProvider urlProvider;
    @inherits Umbraco.Web.Common.AspNetCore.UmbracoViewPage
    @{
	    Layout = null;
        var root = Model.Root();
    }

    <ul>
        <li><a href="@root.Url(urlProvider)">@root.Name</a></li>
        @foreach (var child in root.Children)
        {
            <li><a href="@child.Url(urlProvider)">@child.Name</a></li>
        }
    </ul>

    <h1>@Model.Name</h1>
   ```
* create a content item and publish it
* the url will be populated, click on it and your site should render
* all published pages should render
